### ISSUE TITLE
@azure/msal-angular does not prompt for Microsoft Graph Profile consent on login (even though MSALAngularConfigFactory is configured as such)

### ISSUE THREAD

**Issue by jimbarrett33** [AzureAD/microsoft-authentication-library-for-js#1810]
## Library
- [X] `msal@1.3.2`
- [X] `@azure/msal-angular@1.0.0`

## Framework
Angular

## Description
App is configured according to the Angular 9 sample code which requests consent for "user.read", Microsoft Graph "Sign in and read user profile". However, upon login, the user is not prompted to consent for this permissions (it does however prompt for "View you basic profile" and "Maintain access to data...").

## Error Message
None. Just doesn't prompt for consent. [Issue 1401](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/1401) proposes a workaround but not an actual fix and that issue has been closed.

## Security
- [ ] Is this issue security related?
No

## Regression
- [ ] Did this behavior work before? 
Not sure but [the docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-spa-acquire-token?tabs=angular#acquire-a-token-with-a-pop-up-window ) seem to indicate that you can prompt the user for MS Graph "User.Read" consent during login.

Version:

## MSAL Configuration
```
export const protectedResourceMap: [string, string[]][] = [
  ['https://graph.microsoft.com/v1.0/me', ['user.read']]
];

const isIE = window.navigator.userAgent.indexOf("MSIE ") > -1 || window.navigator.userAgent.indexOf("Trident/") > -1;

function MSALConfigFactory(): Configuration {
  return {
    auth: {
      clientId: '6226576d-37e9-49eb-b201-ec1eeb0029b6',
      authority: "https://login.microsoftonline.com/common/",
      validateAuthority: true,
      redirectUri: "http://localhost:4200/",
      postLogoutRedirectUri: "http://localhost:4200/",
      navigateToLoginRequestUrl: true,
    },
    cache: {
      cacheLocation: "localStorage",
      storeAuthStateInCookie: isIE, // set to true for IE 11
    },
  };
}

function MSALAngularConfigFactory(): MsalAngularConfiguration {
  return {
    popUp: !isIE,
    consentScopes: [
      "user.read",
      "openid",
      "profile"
    ],
    unprotectedResources: ["https://www.microsoft.com/en-us/"],
    protectedResourceMap,
    extraQueryParameters: {}
  };
}
```

## Reproduction steps

1. clone the msal repo/Angular 9 sample https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/angular9-sample-app
2. npm install
3. configure app.module to point to an existing SPA Azure AD app or a newly created app
4. ng serve
5. login as regular user or tenant admin and observe that "user.read" permission is not in the consent list.

## Expected behavior
The "user.read", Microsoft Graph "Sign in and read user profile" consent should be in the list of permissions that the user needs to consent to.

## Browsers/Environment
- [X] Chrome
- [X] Firefox
- [x] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

**Comment by jmckennon** [comment]
This is intended behavior. #1401 moreso addressed a usage problem. From the [Microsoft docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#requesting-individual-user-consent):

>At this time, the offline_access ("Maintain access to data you have given it access to") and user.read ("Sign you in and read your profile") permissions are automatically included in the initial consent to an application. These permissions are generally required for proper app functionality - offline_access gives the app access to refresh tokens, critical for native and web apps, while user.read gives access to the sub claim, allowing the client or app to correctly identify the user over time and access rudimentary user information.

**Comment by jimbarrett33** [comment]
@jmckennon I agree that it is intended to have `user.read` automatically included in the initial consent according to the docs you referred to but it is not happening. Issue [1401](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/1401#issue-585187166) proves this because of the error that occurs when navigating to the profile in the sample app. This means that `user.read` **was not** included as part of the initial consents when logging in. If it was included then the error would not occur.

I'm not sure but maybe the docs are wrong now (that is an very old picture of the Azure login screen in the docs). When I setup an Azure AD app and don't request any special permissions using msal-angular, the default consents are in the picture below, which are `profile` and `offline_access` scopes: 

![image](https://user-images.githubusercontent.com/8396655/85894361-96d06680-b7c2-11ea-93e0-758f0a678b76.png)

It would seem that `user.read`, which equates to "Sign in and read user profile", may be too aggressive for a default consent and it was removed. Just guessing here, but I do know that it is not part of the default consents.

Given the above, my original issue still stands -- specifically requesting `user.read` in the msal-angular MSALAngularConfigFactory consentScopes list **does not prompt for the "Sign in and read user profile"** consent.

**Comment by jmckennon** [comment]
We're looking a bit deeper into this with the server team to get clarity on when some of these nuances occur. In the meantime, can you verify that you have not configured your app to automatically [grant consent to users on an admin's behalf](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#requesting-consent-for-an-entire-tenant)?

**Comment by jimbarrett33** [comment]
Thanks. And, to your question...no, I did not grant admin consent to the app for the entire tenant/all users.

**Comment by jmckennon** [comment]
@jimbarrett33, can you go the Azure Portal -> App Registration -> Click on your app -> API Permissions. There should be a list of permissions for the app, and on the far right you should see some columns like "Status". This should show if you've consented to `user.read` in your app. You can manually delete this consent and run through your app again to see if anything changes.

Additionally, can you check to see if you're getting the `user.read` back from the server when you request it in your scopes?

**Comment by jimbarrett33** [comment]
@jmckennon Admin consent was not enabled in the Azure Portal and user.read is requested. Just in case I started over with these steps:

I created a brand new Azure App registration using the instructions here 
https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-spa-app-registration#create-the-app-registration and using the implicit flow for SPA app. https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-spa-app-registration#create-the-app-registration. The resulting API permissions are here:

![image](https://user-images.githubusercontent.com/8396655/86679845-f4875e80-bfcb-11ea-9a7a-60e96bd7ba18.png)

Notice there is no admin consent consent.

I also used a brand new Angular 9 sample app from here https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-angular-samples/angular9-sample-app and replaced the clientId in the app.module.ts (line 34) to be the client ID of my new "spa test" app (and also commented out the special api permission for the To Do app in the sample (line 55). 

When I sign-in as a non-admin user I get this consent screen:
![image](https://user-images.githubusercontent.com/8396655/86680778-eab22b00-bfcc-11ea-97c9-0d60e5d88d8b.png)

This is all successful.

When I then navigate to profile screen and the Angular auth guard activates, this silent request is sent out (I left some url params off):

`https://login.microsoftonline.com/<my-tenant-id>/oauth2/v2.0/authorize?response_type=token
&scope=user.read%20openid%20profile&client_id=0391e74f-4432-49fb-88c8-31d79db68640`

Notice that user.read is in the request.

This results in the following error written to the console by the MSAL logger.

```
MSAL Logging:  Tue, 07 Jul 2020 00:51:46 GMT:a4fecf6c-a0b3-41d5-87ce-b2c59ae58f4a-1.3.2-Error Error when acquiring token for scopes: user.read InteractionRequiredAuthError: AADSTS65001: The user or administrator has not consented to use the application with ID '0391e74f-4432-49fb-88c8-31d79db68640' named 'spa test'. Send an interactive authorization request for this user and resource.
Trace ID: d556bf43-1964-4d32-becc-8b7f85d38b00
Correlation ID: 639e9bb5-41ef-4540-8f49-b53bf5a40a3f
Timestamp: 2020-07-07 00:51:47Z
```

Hope this helps.

**Comment by tnorling** [comment]
@jimbarrett33 Can you confirm that `User.Read` was also included in the initial login request? It is expected that `acquireTokenSilent` would fail if you haven't first included the scope in an interactive request and consented to it. The only scopes Msal requests by default if none are provided are `openid` and `profile`

**Comment by jimbarrett33** [comment]
@tnorling user.read is NOT included in the initial login request. That is really the crux of this issue. If I supply the following configuration MsalAngularConfiguration, `user.read` is not included in the scopes request.

```
function MSALAngularConfigFactory(): MsalAngularConfiguration {
  return {
    popUp: !isIE,
    consentScopes: [
      "user.read",
      "openid",
      "profile"
    ],
    unprotectedResources: ["https://www.microsoft.com/en-us/"],
    protectedResourceMap,
    extraQueryParameters: {}
  };
}
```

here is the url request (with some params omitted for brevity):
```
https://login.microsoftonline.com/common/oauth2/v2.0/authorize
?response_type=id_token
&scope=openid%20profile
&client_id=0391e74f-4432-49fb-88c8-31d79db68640
```

**Comment by technical-boy** [comment]
Hi @jimbarrett33 . The consent scopes provided in your configuration will only be included in login requests that are initiated by the MSAL Guard. If you are invoking `loginPopup` or `loginRedirect` manually in your application, you will need to provide any additional scopes explicitly.

```js
this.authService.loginRedirect({
  additionalScopesToConsent: ['user.read']
});
```

Could you try this and let us know if it fixes your issue? Thanks!

**Comment by jimbarrett33** [comment]
Hi @technical-boy. I tried your suggestion and this did not prompt for the additional consent. In my version, I think the latest, the option is actually called `extraScopesToConsent`

```
      this.authService.loginPopup( {
        extraScopesToConsent: ['user.read']
      } );
```

Seems it just silently fails -- I didn't notice anything bad in the console.

**Comment by sudipbanfse** [comment]
facing similar issue. Trying to implement on behalf of flow with implicit grant flow. So SPA calls backend api and backend api in turns calls microsoft graph api.  
Configuration Details:
1. SPA has given API permission to backend api as delegated scope without admin consent in app registration.
2. BAckend API adds SPA as known client application and exposes api with scope , which can be granted by user and admin.
3. Backend api has added Microsoft graph user.read and sites.manager.all as delegated consent without admin consent.
4. All configurations are done as per documentation and mentioned above by @jimbarrett33 .
5. Only catch - using Angular latest version (10.x lts).
6. SPA is not using MSAL guard as of now, manually calling loginpopup() from start up (app.component.ts)

Ideally :
1. The SPA should display combined consent but as mentioned by @jimbarrett33  only displaying default consent.

Issue:
Though access token is generated and sent to backend api, while  backend api calling MS graph API getting 

- [ ] [MsalUiRequiredException](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.msaluirequiredexception?view=azure-dotnet) UI intervention required error message. Though intention is to provide it upfront as part of user login as per on behalf flow documentation.

**Comment by jmckennon** [comment]
@sudipbanfse Can you open a new issue with you description here and any other msal configurations and versions? Thanks!

**Comment by jimbarrett33** [comment]
I am leaving a comment since I don't want this issue to auto-close. I believe there is something to fix here.

**Comment by roddev-v** [comment]
Having the same problem with the permissions for the calendar API. The login prompt doesn't ask for the calendar permissions in msal 1.4.0 / azure-msal 1.1.0

**Comment by tnorling** [comment]
@jimbarrett33 Apologies for the delay on this, I had some time to repro this and here's what I found. If you're using the sample as-is the login calls do not provide any scopes, which would explain why you are not being asked for consent on login. If you go to `app.component.ts` you'll find a `login()` function which in turn calls either `loginRedirect` or `loginPopup`, if you add your scopes here do you get the expected behavior? 

As mentioned previously by my colleague, the `consentScopes` config option is only used for the Guard so if you're using the Guard and are not seeing the correct scopes in the request we should take a look at that. Likewise if you are explicitly providing scopes to `loginPopup` or `loginRedirect` and are not seeing those scopes in the request we should understand why, after adding the scopes to the login function described above I was able to get the User.Read consent screen. Let me know if that helps clear things up!

**Comment by tnorling** [comment]
Closing due to inactivity and do not believe this is a bug. If the above suggestions do not resolve the issue please open a new issue so we can follow up.

**Comment by jimbarrett33** [comment]
Thanks @tnorling. I will take a look at this soon. Appreciate your time and effort in looking into it.

**Comment by tomatsue** [comment]
FYI. I had the same issue. Following this thread and the docs, I added `extraScopesToConsent` to `loginPopup` but it didn't work. It seems because `extraScopesToConsent`  won't be appended the scopes when the `scopes` is null. 
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/2e93cfe4df6a2ef6c6f53f540ba0b84cc6746781/lib/msal-core/src/ScopeSet.ts#L145

So I needed to add `scopes` to `loginPopup` instead of `extraScopesToConsent`.
```javascript
      this.authService.loginPopup({
        scopes: ["api://<webapi-id>/access_as_user"]
      });
```
